luatest/diff.lua

```diff
@@ -0,0 +1,218 @@
+            for i = 1, #child do
+                lines[#lines + 1] = child[i]
+            end
+        end
+    end
+
+    lines[#lines + 1] = indent .. '}'
+    return lines
+end
+
+--- Convert a supported Lua value into a textual form suitable for diffing.
+--
+-- * Tables are serialized with recursive normalization and a simple textual encoder.
+-- * Strings are used as-is.
+-- * Numbers / booleans are converted via tostring().
+-- * Top-level opaque userdata/cdata disable diffing when tostring() fails (return nil).
+local function as_text(value)
+    local t = type(value)
+
+    if t == 'cdata' or t == 'userdata' then
+        -- Top-level: if we can't get a meaningful tostring(), disable diffing.
+        local ok, s = pcall(tostring, value)
+        if ok and type(s) == 'string' then
+            return s
+        end
+        return nil
+    end
+
+    if t == 'string' then
+        return value
+    end
+
+    local normalized = normalize_for_text(value)
+    return table.concat(encode_value(normalized), '\n')
+end
+
+local function diff_by_lines(text1, text2)
+    local lines1 = string.split(text1, '\n')
+    local lines2 = string.split(text2, '\n')
+
+    local m = #lines1
+    local n = #lines2
+    local lcs = {}
+
+    for i = 0, m do
+        lcs[i] = {}
+        lcs[i][0] = 0
+    end
+
+    for j = 0, n do
+        lcs[0][j] = 0
+    end
+
+    for i = 1, m do
+        for j = 1, n do
+            if lines1[i] == lines2[j] then
+                lcs[i][j] = lcs[i - 1][j - 1] + 1
+            else
+                local left = lcs[i - 1][j]
+                local top = lcs[i][j - 1]
+                lcs[i][j] = left >= top and left or top
+            end
+        end
+    end
+
+    local diffs = {}
+    local i = m
+    local j = n
+
+    while i > 0 or j > 0 do
+        if i > 0 and j > 0 and lines1[i] == lines2[j] then
+            table.insert(diffs, 1, {'equal', lines1[i]})
+            i = i - 1
+            j = j - 1
+        elseif j > 0 and (i == 0 or lcs[i][j - 1] >= lcs[i - 1][j]) then
+            table.insert(diffs, 1, {'insert', lines2[j]})
+            j = j - 1
+        else
+            table.insert(diffs, 1, {'delete', lines1[i]})
+            i = i - 1
+        end
+    end
+
+    return diffs
+end
+
+local function prettify_patch(diffs)
+    local out = {}
+
+    for _, diff in ipairs(diffs) do
+        local tag, line = diff[1], diff[2]
+        if tag == 'equal' then
+            table.insert(out, line)
+        elseif tag == 'delete' then
+            table.insert(out, '-' .. line)
+        elseif tag == 'insert' then
+            table.insert(out, '+' .. line)
+        end
+    end
+
+    if #out == 0 then
+        return nil
+    end
+
+    return table.concat(out, '\n')
+end
+
+--- Build unified diff for expected and actual values serialized to text.
+-- Returns nil when values can't be serialized or there is no diff.
+function M.build_unified_diff(expected, actual)
```

Comment:
The function name is misleading because the produced diff doesn't conform to the unified format.